src/execution_plan/execution_plan.c

```diff
@@ -527,14 +552,105 @@ static inline void _buildUnwindOp(ExecutionPlan *plan, const cypher_astnode_t *c
 	_ExecutionPlan_UpdateRoot(plan, op);
 }
 
-static inline void _buildMergeOp(GraphContext *gc, AST *ast, ExecutionPlan *plan,
-								 const cypher_astnode_t *clause, ResultSetStatistics *stats) {
+// TODO choose place and names for these functions
+static const char **_GetPathVariables(const AST *ast, const cypher_astnode_t *path) {
+	uint count = cypher_ast_pattern_path_nelements(path);
+	const char **variables = array_new(const char *, count);
+
+	for(uint i = 0; i < count; i ++) {
+		const cypher_astnode_t *elem = cypher_ast_pattern_path_get_element(path, i);
+		const char *alias = AST_GetEntityName(ast, elem);
+		variables = array_append(variables, alias);
+	}
+
+	return variables;
+}
+
+static const char **_MergeCollectVariables(const AST *ast, const cypher_astnode_t *clause) {
+	const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(clause);
+	return _GetPathVariables(ast, path);
+}
+
+static void _buildMergeRightHandStreams(AST *ast, ExecutionPlan *plan,
+										const cypher_astnode_t *clause, ResultSetStatistics *stats, OpBase *merge_op) {
+	// Initialize an ExecutionPlan that shares this plan's Record mapping.
+	ExecutionPlan *rhs_plan = _NewEmptyExecutionPlan();
+	rhs_plan->record_map = plan->record_map;
+	const char **variables = NULL;
+	// If we have previous ops, determine which of this clause's variables are bound (if any)
+	// and build an Argument op as the source operation of the new ExecutionPlan.
+	if(merge_op->children[0]) {
+		variables = _MergeCollectVariables(ast, clause);
+		uint variable_count = array_len(variables);
+		rax *bound_variables = raxNew();
+		ExecutionPlan_ResolvedModifiers(plan->root, bound_variables);
+
+		for(uint i = 0; i < variable_count; i++) {
+			if(raxFind(bound_variables, (unsigned char *)variables[i], strlen(variables[i])) == raxNotFound) {
+				// Variable not found, delete from array
+				variables = array_del(variables, i);
+				variable_count--;
+				i--; // Deleted the current variable, repeat check at this position.
+			}
+		}
+
+		// If we have bound variables, build an Argument op that represents them.
+		if(variable_count > 0) rhs_plan->root = NewArgumentOp(plan, variables);
+	}
+
+	// Build a temporary AST holding the MERGE path within a MATCH clause.
+	const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(clause);
+	AST *rhs_ast = AST_MockMatchPattern(ast, path);
+
+	_NewExecutionPlan(rhs_plan, NULL);
+
+	AST_Free(rhs_ast);
```

Comment:
Memory leak on the constructed MATCH node currently, and I'm a little nervous about whether we'll encounter problems when fixing that leak. Cloning the MATCH pattern strips it of AST  annotations, freeing it will double-free those annotations, etc.

I'd rather get the core logic in place first and worry about those parts after!